### PR TITLE
 Ad admin access

### PR DESCRIPTION
**Pull request by apalski**
PR to check for admin status for users index view - only a user with admin == true can see the index view, the full list of registered users.
- add admin? and require_admin methods to application controller
- add before_action require_admin for index action
- add feature specs to test admin access to index view
- add admin to user_params in users controller
- add admin field to new view with condition that current_user must be admin for the field to show
- add translations for admin access alerts
- refactor user routes to include index route
- add spec helper methods log_in_as and feature_log_in_as
Looking for feedback on all additions and changes, particularly anything I have missed or situations not covered in specs.
https://trello.com/c/DHCvfuJH
https://trello.com/c/vtH5Zsdk
Thank you for looking at this for me!

### CODE REVIEW

File: app/views/users/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<div><%= t(".title") %></div>
+<ul><% @users.each do |user| %>
+  <li><%= user.name %></li>
```
Comment (scara82) [inline app/views/users/index.html.erb]: I have just one small comment on the style, I have no idea what elle thinks about that, but I prefer have the `<ul>` alone and the erb next line, I think that it's just easer to read. 😄 
```
<ul>
  <%= ....   %>
```

File: spec/controllers/users_controller_spec.rb [hunk 2]
```diff
@@ -48,6 +49,30 @@
           to match I18n.t("flash.actions.create.alert", resource_name: "User")
       end
     end
+
+    context "current user not admin" do
+      it "it won't create a new user with admin param" do
```
Comment (scara82) [inline spec/controllers/users_controller_spec.rb]: Double `it`... :mag_right:

File: app/views/users/index.html.erb (hunk 1 above)
Comment (apalski) [reply app/views/users/index.html.erb]: Thanks Marta, that's the way I would normally do it too!

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -11,7 +11,18 @@ def current_user
...
+  def require_admin
+    unless admin?
+      redirect_to user_path(current_user), alert: I18n.t(
+        "users.index.require_admin"
+      )
+    end
+  end
+
+  def admin?
+    current_user.admin == true
```
Comment (elle) [inline app/controllers/application_controller.rb]: This can change to `current_user.admin?`

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,4 +1,12 @@
 class UsersController < ApplicationController
+  before_action :require_admin, only: [:index]
+
+  def index
+    @users = User.all
+
+    respond_with(@users)
+  end
```
Comment (elle) [inline app/controllers/users_controller.rb]: I would consider using admin namespace in routes, and have admin controllers for that. We can chat more about this next week. Maybe for another PR to refactor this out.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -30,7 +38,7 @@ def destroy
   private
 
   def user_params
-    params.require(:user).permit(:name, :email, :password, :uid)
+    params.require(:user).permit(:name, :email, :password, :admin, :uid)
```
Comment (elle) [inline app/controllers/users_controller.rb]: I would suggest not allowing `:admin` to be passed as a param. Because then every user could hack the system and set themselves as admin.

Once again, I would suggest allowing this only under an admin namespaced in admin controllers.

File: app/views/users/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<div><%= t(".title") %></div>
```
Comment (elle) [inline app/views/users/index.html.erb]: Since this seems to be a title, it makes sense (and more semantic) to use a heading tag, rather than a generic div tag. Using header tags gives these element more value when bots inspect the page for SEO

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -1,7 +1,10 @@
-<h1><%= t(".title") %></h1>
+<div><%= t(".title") %></div>
```
Comment (elle) [inline app/views/users/new.html.erb]: Same comment about the heading

File: app/views/users/edit.html.erb [hunk 8]
```diff
@@ -1,4 +1,4 @@
-<h1><%= t(".title", name: user.name) %></h1>
+<div><%= t(".title") %></div>
```
Comment (elle) [inline app/views/users/edit.html.erb]: Same comment about the heading

File: app/views/users/new.html.erb [hunk 9]
```diff
@@ -1,7 +1,10 @@
...
 <%= simple_form_for @user do |f| %>
   <%= f.input :name %>
   <%= f.input :email %>
   <%= f.input :password %>
+  <% if current_user&.admin? %>
```
Comment (elle) [inline app/views/users/new.html.erb]: This line `if current_user&.admin?` hides that it really is about `if current_user and current_user.admin?`
Since you already have an `admin?` helper method, I would use that instead.

Also, once more, consider creating admin namespaced controllers for this :)

File: db/schema.rb [hunk 10]
```diff
@@ -34,6 +32,16 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "goals", force: :cascade do |t|
```
Comment (elle) [inline db/schema.rb]: This table does not belong in this PR. Try and think of it as keeping each commit as a fully working application.
In this case, there will be a `goals` table without any other application support for its existence.

File: spec/features/users_creates_new_account_spec.rb [hunk 11]
```diff
@@ -1,5 +1,28 @@
 require "rails_helper"
 
+feature "admin user accesses #index" do
```
Comment (elle) [inline spec/features/users_creates_new_account_spec.rb]: This scenario fits in a new file: `spec/features/admin_accesses_users_list_spec.rb`

File: spec/features/users_creates_new_account_spec.rb [hunk 12]
```diff
@@ -1,5 +1,28 @@
...
+    user = create(:user, admin: true)
+    feature_log_in_as user
+
+    visit users_path
+
+    expect(page).to have_text(I18n.t("users.index.title"))
+  end
+end
+
+feature "non-admin user accesses #index" do
```
Comment (elle) [inline spec/features/users_creates_new_account_spec.rb]: This should be moved to a controller spec

File: spec/features/users_creates_new_account_spec.rb [hunk 13]
```diff
@@ -17,42 +40,72 @@
       ))
   end
 
-  scenario "views show page" do
-    user = create(:user)
+  scenario "successfully as admin" do
```
Comment (elle) [inline spec/features/users_creates_new_account_spec.rb]: This scenario lives in a file that is called `users_create_new_account_spec.rb` (really it should be singular: `user_creates_new_account_spec.rb`)

It should move to its own file.
Also, I would do different feature spec files for admin: for example: `admin_can_set_another_user_as_admin_spec.rb`.

Comment (elle) [reply spec/features/users_creates_new_account_spec.rb]: This spec file has too much in it.
```
feature "admin user accesses #index" do
  scenario "successfully" do
  end
end

feature "non-admin user accesses #index" do
  scenario "unsuccessfully" do
  end
end

feature "user creates new account" do
  scenario "successfully" do
  end

  scenario "successfully as admin" do
  end

  feature "views show page" do
    scenario "successfully" do
    end
  end

  feature "updates profile" do
    scenario "successfully" do
    end
  end

  feature "deletes account" do
    scenario "successfully" do
  end
end
```

Each one of these should be its own file:
- user creates an account
- user updates their details
- user delates their account
- admin views all users
- admin can set another user as admin

Comment (elle) [inline spec/features/users_creates_new_account_spec.rb]: What does `scenario "successfully as admin" do` this scenario do?

File: spec/helpers/helpers.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
+module Helpers
+  def log_in_as(user, options = {})
```
Comment (elle) [inline spec/helpers/helpers.rb]: Discussed this file in the other PR

File: spec/rails_helper.rb [hunk 15]
```diff
@@ -4,6 +4,7 @@
 abort("DATABASE_URL environment variable is set") if ENV["DATABASE_URL"]
 
 require "rspec/rails"
+require "spec_helper"
```
Comment (elle) [inline spec/rails_helper.rb]: Comment in other PR

File: spec/spec_helper.rb [hunk 16]
```diff
@@ -11,6 +11,7 @@
 
 require "webmock/rspec"
 require "timecop"
+require "helpers/helpers"
```
Comment (elle) [inline spec/spec_helper.rb]: Comment in other PR

File: spec/spec_helper.rb [hunk 17]
```diff
@@ -25,6 +26,8 @@
 
   config.example_status_persistence_file_path = "tmp/rspec_examples.txt"
   config.order = :random
+
+  config.include Helpers
```
Comment (elle) [inline spec/spec_helper.rb]: Comment in other PR

File: app/views/users/show.html.erb [hunk 18]
```diff
@@ -1,2 +1,2 @@
-<h1><%= t(".welcome", name: user.name) %></h1>
-<h2><%= t(".profile_information", email: user.email) %></h2>
+<div><%= t(".welcome") %></div>
```
Comment (elle) [inline app/views/users/show.html.erb]: I would prefer to use a heading tag here.

File: config/locales/en.yml [hunk 19]
```diff
@@ -56,14 +56,17 @@ en:
...
     new:
       title: "Please fill in the form to sign up"
     edit:
       delete: "Delete Account"
-      title: "%{name}, please fill in the form to edit your profile"
+      title: "Please fill in the form to edit your profile"
     show:
       profile_information: "Your email address is: %{email}"
-      welcome: "Welcome %{name}"
+      welcome: "Welcome your details are:"
```
Comment (elle) [inline config/locales/en.yml]: Add comma: `Welcome, your details are:`

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (apalski) [reply app/controllers/users_controller.rb]: OK, sounds good, I am interested to give this a go

File: app/views/users/new.html.erb (hunk 7 above)
Comment (apalski) [reply app/views/users/new.html.erb]: As above, done sha: 1ea0d5e

File: app/views/users/index.html.erb (hunk 6 above)
Comment (apalski) [reply app/views/users/index.html.erb]: Cool, good to know! Done in sha: 1ea0d5e

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (apalski) [reply app/controllers/users_controller.rb]: Good point, sounds like admin namespacing for the routes makes sense, done in sha: 1ea0d5e

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (apalski) [reply app/controllers/application_controller.rb]: duh! of course, thank you done in sha: 1ea0d5e

File: db/schema.rb (hunk 10 above)
Comment (apalski) [reply db/schema.rb]: I don't know why this happens, I have a goals model branch which has this migration and table in the schema. For some reason if tables and migrations exist in other branches they seem to appear in a new branch. I didn't create the goals migration in this branch???

File: spec/features/users_creates_new_account_spec.rb (hunk 12 above)
Comment (apalski) [reply spec/features/users_creates_new_account_spec.rb]: I don't seem to be able to access methods in the application controller from the controller specs?

File: app/views/users/show.html.erb (hunk 18 above)
Comment (apalski) [reply app/views/users/show.html.erb]: Done ins sha: 1ea0d5e

File: config/locales/en.yml (hunk 19 above)
Comment (apalski) [reply config/locales/en.yml]: Done in sha: 1ea0d5e

File: spec/features/users_creates_new_account_spec.rb (hunk 11 above)
Comment (apalski) [reply spec/features/users_creates_new_account_spec.rb]: Done in sha: b2bcac4

File: spec/features/users_creates_new_account_spec.rb (hunk 13 above)
Comment (apalski) [reply spec/features/users_creates_new_account_spec.rb]: Done in sha: b2bcac4

Comment (apalski) [reply spec/features/users_creates_new_account_spec.rb]: Changed to admin creates new admin account. Done in sha: b2bcac4

**Comment by apalski** [comment]
Ready to merge, thanks for the feedback @elle and @scara82 !
